crates/assists/src/utils/insert_use.rs

```diff
@@ -22,525 +18,566 @@ pub(crate) fn find_insert_use_container(
+    // track the last element we iterated over, if this is still None after the iteration then that means we never iterated in the first place
+    let mut last = None;
+    // find the element that would come directly after our new import
+    let post_insert =
+        group_iter.inspect(|(_, node)| last = Some(node.clone())).find(|(path, _)| {
+            let check_segments = segment_iter(&path);
+            segments
+                .clone()
+                .zip(check_segments)
+                .flat_map(|(seg, seg2)| seg.name_ref().zip(seg2.name_ref()))
+                .all(|(l, r)| l.text() <= r.text())
+        });
+    match post_insert {
+        // insert our import before that element
+        Some((_, node)) => (InsertPosition::Before(node.into()), AddBlankLine::After),
+        // there is no element after our new import, so append it to the end of the group
+        None => match last {
+            Some(node) => (InsertPosition::After(node.into()), AddBlankLine::Before),
+            // the group we were looking for actually doesnt exist, so insert
+            None => {
+                // similar concept here to the `last` from above
+                let mut last = None;
+                // find the group that comes after where we want to insert
+                let post_group = path_node_iter
+                    .inspect(|(_, node)| last = Some(node.clone()))
+                    .find(|(p, _)| ImportGroup::new(p) > group);
+                match post_group {
+                    Some((_, node)) => {
+                        (InsertPosition::Before(node.into()), AddBlankLine::AfterTwice)
+                    }
+                    // there is no such group, so append after the last one
+                    None => match last {
+                        Some(node) => (InsertPosition::After(node.into()), AddBlankLine::Before),
+                        // there are no imports in this file at all
+                        None => (InsertPosition::First, AddBlankLine::AfterTwice),
+                    },
+                }
+            }
+        },
     }
+}
 
-    fn add_in_tree_list(
-        common_segments: usize,
-        tree_list: ast::UseTreeList,
-        add_self: bool,
-    ) -> Self {
-        ImportAction::AddInTreeList { common_segments, tree_list, add_self }
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use test_utils::assert_eq_text;
+
+    #[test]
+    fn insert_start() {
+        check_none(
+            "std::bar::A",
+            r"use std::bar::B;
+use std::bar::D;
+use std::bar::F;
+use std::bar::G;",
+            r"use std::bar::A;
+use std::bar::B;
+use std::bar::D;
+use std::bar::F;
+use std::bar::G;",
+        )
     }
 
-    fn better(left: ImportAction, right: ImportAction) -> ImportAction {
-        if left.is_better(&right) {
-            left
-        } else {
-            right
-        }
+    #[test]
+    fn insert_middle() {
+        check_none(
+            "std::bar::E",
```

Comment:
Maybe call this and other inserted imports as `::EE` or something like that, such that they better stand out in tests?